File: 25-ReverseNodesink-Group/25-ReverseNodesink-Group.java
/*

25. Reverse Nodes in k-Group
Hard

Given a linked list, reverse the nodes of a linked list k at a time and return its modified list.

k is a positive integer and is less than or equal to the length of the linked list. If the number of nodes is not a multiple of k then left-out nodes in the end should remain as it is.

Example:

Given this linked list: 1->2->3->4->5

For k = 2, you should return: 2->1->4->3->5

For k = 3, you should return: 3->2->1->4->5

Note:

    Only constant extra memory is allowed.
    You may not alter the values in the list's nodes, only nodes itself may be changed.

*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null) return head;
        Stack<ListNode> s = new Stack<>();
        ListNode pseudoHead = new ListNode(0);
        ListNode pseudotemp = pseudoHead;
        ListNode temp = head;
        int len = getLength(temp, k);
        for (int i = 0; i< len/k; i++){
            int count = 0;
            while(count++ != k){
                ListNode t = head.next;
                s.push(head);
                head.next = null;
                head = t;
            }
            while(!s.isEmpty()){
                pseudotemp.next = s.pop();
                pseudotemp = pseudotemp.next;
            }
        }
        pseudotemp.next = head;
        return pseudoHead.next;
        
    }
    
    private int getLength(ListNode temp, int k){
        int count = 0;
        while(temp != null){
            count++;
            temp = temp.next;
        }
        return count;
    }
}

File: 25-ReverseNodesink-Group/25-ReverseNodesink-Group.py
"""

25. Reverse Nodes in k-Group

Given the head of a linked list, reverse the nodes of the list k at a time, and return the modified list.

k is a positive integer and is less than or equal to the length of the linked list. If the number of nodes is not a multiple of k then left-out nodes, in the end, should remain as it is.

You may not alter the values in the list's nodes, only nodes themselves may be changed.

Example 1:

Input: head = [1,2,3,4,5], k = 2
Output: [2,1,4,3,5]

Example 2:

Input: head = [1,2,3,4,5], k = 3
Output: [3,2,1,4,5]

Constraints:

    The number of nodes in the list is n.
    1 <= k <= n <= 5000
    0 <= Node.val <= 1000

 

Follow-up: Can you solve the problem in O(1) extra memory space?

"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def get_length(head):
            count = 0
            while head:
                count += 1
                head = head.next
            return count
        temp = head
        length = get_length(temp)
        newHead = ListNode(0)
        temp = newHead
        q = deque()
        for i in range(length//k):
            for j in range(k):
                q.append(head)
                head = head.next
            while q:
                temp.next = ListNode(q.pop().val)
                temp = temp.next
        temp.next = head
        return newHead.next
